moq-karp/src/broadcast.rs

```diff
@@ -1,116 +1,175 @@
-use crate::{Audio, Catalog, Error, Result, Track, TrackConsumer, TrackProducer, Video};
+use crate::{Audio, Catalog, Error, Result, Track, TrackProducer, Video};
 
 use moq_async::{spawn, Lock};
 use moq_transfork::{Announced, AnnouncedConsumer, AnnouncedMatch, Session};
+use crate::track::TrackConsumer;
 
 use derive_more::Debug;
 
+#[derive(Clone)]
+struct BroadcastListener {
+	pub session: Session,
+	pub catalog: Lock<CatalogProducer>,
+}
+
+#[derive(Debug, Clone)]
+struct Context {
+	pub video: Option<Video>,
+	pub audio: Option<Audio>,
+	pub video_consumer: Option<moq_transfork::TrackConsumer>,
+	pub audio_consumer: Option<moq_transfork::TrackConsumer>,
+}
+impl Context {
+	pub fn default() -> Self {
+		Self {
+			video: None,
+			audio: None,
+			video_consumer: None,
+			audio_consumer: None,
+		}
+	}
+
+	fn publish_video_to(&self, listener: &mut BroadcastListener) -> anyhow::Result<()> {
+		if let Some(consumer) = self.video_consumer.clone() {
+			if let Some(info) = self.video.clone() {
+				let mut catalog = listener.catalog.lock();
+				catalog.current.video.push(info.clone());
+				catalog.publish()?;
+
+				// let consumer = producer.subscribe();
+				let consumer = TrackConsumer::new(consumer);
+				listener.session.publish(consumer.track.clone())?;
+
+				// Start a task that will remove the catalog on drop.
+				let catalog = listener.catalog.clone();
+				let track_info = info.track;
+				spawn(async move {
+					consumer.closed().await.ok();
+
+					let mut catalog = catalog.lock();
+					catalog.current.video.retain(|v| v.track != track_info);
+					catalog.publish().unwrap();
+				});
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn publish_audio_to(&self, listener: &mut BroadcastListener) -> anyhow::Result<()> {
+		if let Some(consumer) = self.audio_consumer.clone() {
+			if let Some(info) = self.audio.clone() {
+				let mut catalog = listener.catalog.lock();
+				catalog.current.audio.push(info.clone());
+				catalog.publish()?;
+
+				// let consumer = producer.subscribe();
+				let consumer = TrackConsumer::new(consumer);
+				listener.session.publish(consumer.track.clone())?;
+
+				// Start a task that will remove the catalog on drop.
+				let catalog = listener.catalog.clone();
+				let track_info = info.track;
+				spawn(async move {
+					consumer.closed().await.ok();
+
+					let mut catalog = catalog.lock();
+					catalog.current.audio.retain(|v| v.track != track_info);
+					catalog.publish().unwrap();
+				});
+			}
+		}
+
+		Ok(())
+	}
+}
+
 #[derive(Debug, Clone)]
 #[debug("{:?}", path)]
 pub struct BroadcastProducer {
-	pub session: Session,
+	listeners: Vec<BroadcastListener>,
```

Comment:
This will behave kinda strange with `Clone`. You either need to remove `Clone`, or put all state behind a `Lock`. I think this is `Clone` because somebody wanted to query the catalog and that was just the easiest way to do it.

IMO:
```rust
#[derive(Debug, Clone)]
struct BroadcastProducerState {
    catalog: CatalogProducer,
    tracks: Vec<TrackProducer>,
    subscribers: Vec<Session>,
}

#[derive(Debug, Clone)]
#[debug("{:?}", path)]
pub struct BroadcastProducer {
  pub path: String,
  id: u64,
  state: Lock<BroadcastProducerState>,
}
```

Reply:
Also, you should have a way to remove a `BroadcasterListener` when the session closes. Otherwise it's a memory leak.

And yeah that would involve a background task unfortunately.

Reply:
An alternative API would be something like `BroadcastProducer/BroadcastConsumer` pair in moq_transfork with a:

```rust
impl BroadcastProducer {
  pub fn publish(&mut self, track: TrackConsumer) {
    // inserts a new track and notifies all consumers
  }
}  

impl BroadcastConsumer {
  pub fn next(&mut self) -> Option<TrackConsumer> {
    // .. waits for a new track
  }
}
```

That way you wouldn't need a Vec of subscribers. You can instead spawn a background task to block on `broadcast.next()` and `session.closed()`.